### PR TITLE
Prevents cycles in cascading workflows

### PR DESCRIPTION
**Pull request by saurav-c**
## Describe your changes and why you are making these changes
This PR prevents cycles in cascading workflows.

## Related issue number (if any)
ENG 2198

Testing: I manually tested to ensure that an error is returned on the SDK when a user forms a cycle. I am currently adding integration test cases for this (and also add unit tests for the graph cycle detection code).

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: src/golang/lib/workflow/validate.go [hunk 1]
```diff
@@ -69,9 +80,53 @@ func ValidateSchedule(
...
+	if hasCycle := checkForCycle(workflowID, schedule.SourceID, cascadingWorkflows); hasCycle {
+		return http.StatusBadRequest, errors.New("Cannot allow cycles for cascading workflows.")
+	}
+
 	return http.StatusOK, nil
 }
+
+// checkForCycle returns true if setting workflowID's source workflow to sourceID would
+// result in a cycle.
+func checkForCycle(workflowID uuid.UUID, sourceID uuid.UUID, targetWorkflows []models.Workflow) bool {
```
Comment (likawind) [inline src/golang/lib/workflow/validate.go:117]: can we have a task to add unit test for this method?

File: src/golang/lib/graph/directed.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+	for len(queue) > 0 {
+		root = queue[0]
+		queue = queue[1:]
+
+		if root == dest {
+			return true
+		}
+
+		queue = append(queue, g.nodes[root].edges...)
```
Comment (likawind) [inline src/golang/lib/graph/directed.go:41]: so this grows fast as we don't de-dupe when we add new nodes to the queue?

I think it works for cascading workflows as the graph is simple but probably not for more complex networks. Given this library is for generic use I would consider adding a task to optimize the algorithm

File: src/golang/lib/workflow/validate.go [hunk 3]
```diff
@@ -7,24 +7,34 @@ import (
...
+// workflowID can be ignored.
+// The following conditions are not allowed:
+// 1. Using a CascadingUpdateTrigger when engineType is Airflow.
+// 2. Having a CascadingUpdateTrigger where SourceID is for a Workflow that is
 // running on a non self-orchestrated engine, such as Airflow. This is not allowed
 // since Aqueduct cannot trigger Workflow runs at the end of execution on an
 // engine that is not self-orchestrated.
+// 3. Having a CascadingUpdateTrigger that creates a cycle amongst the cascading workflows.
 // It returns an HTTP status code and a client-friendly error, if any.
 func ValidateSchedule(
```
Comment (likawind) [inline src/golang/lib/workflow/validate.go:34]: just confirming that this function is called within the txn during workflow registration?

File: src/golang/lib/workflow/validate.go (hunk 1 above)
Comment (likawind) [reply src/golang/lib/workflow/validate.go:117]: saw that in PR comments, thanks!

**Review by likawind — APPROVED** [review]

File: src/golang/lib/graph/directed.go (hunk 2 above)
Comment (saurav-c) [reply src/golang/lib/graph/directed.go:41]: https://linear.app/aqueducthq/issue/ENG-2204/optimize-libgraph-haspath-function-for-directedgraph

File: src/golang/lib/workflow/validate.go (hunk 3 above)
Comment (saurav-c) [reply src/golang/lib/workflow/validate.go:34]: Good catch, I moved it inside the txn

**Pull request merged**